File: Tests/AINotebookCoreTests/OllamaPullEventTests.swift
import XCTest
@testable import AINotebookCore

final class OllamaPullEventTests: XCTestCase {
    func testDecodeStartStatus() throws {
        let json = #"{"status":"pulling manifest"}"#.data(using: .utf8)!
        let event = try JSONDecoder().decode(OllamaPullEvent.self, from: json)
        XCTAssertEqual(event.status, "pulling manifest")
        XCTAssertNil(event.total)
        XCTAssertNil(event.completed)
        XCTAssertNil(event.digest)
    }

    func testDecodeProgressEvent() throws {
        let json = """
        {"status":"downloading","digest":"sha256:abc","total":2019377664,"completed":1000000}
        """.data(using: .utf8)!
        let event = try JSONDecoder().decode(OllamaPullEvent.self, from: json)
        XCTAssertEqual(event.status, "downloading")
        XCTAssertEqual(event.digest, "sha256:abc")
        XCTAssertEqual(event.total, 2_019_377_664)
        XCTAssertEqual(event.completed, 1_000_000)
        XCTAssertEqual(event.fractionComplete!, 1_000_000.0 / 2_019_377_664.0, accuracy: 1e-9)
    }

    func testFractionCompleteIsNilWhenMissing() throws {
        let json = #"{"status":"verifying"}"#.data(using: .utf8)!
        let event = try JSONDecoder().decode(OllamaPullEvent.self, from: json)
        XCTAssertNil(event.fractionComplete)
    }

    func testIsTerminalSuccess() {
        let success = OllamaPullEvent(status: "success", digest: nil, total: nil, completed: nil)
        XCTAssertTrue(success.isTerminalSuccess)

        let mid = OllamaPullEvent(status: "downloading", digest: nil, total: 100, completed: 50)
        XCTAssertFalse(mid.isTerminalSuccess)
    }
}

File: windows/tests/AINotebook.Core.Tests/Ollama/OllamaPullEventTests.cs
using System.Text.Json;
using AINotebook.Core.Ollama;
using Xunit;

namespace AINotebook.Core.Tests.Ollama;

public class OllamaPullEventTests
{
    private static readonly JsonSerializerOptions Opts = OllamaJson.Options;

    // OllamaPullEventTests.testDecodeStartStatus
    [Fact]
    public void DecodeStartStatusLeavesProgressNull()
    {
        var ev = JsonSerializer.Deserialize<OllamaPullEvent>("""{"status":"pulling manifest"}""", Opts)!;
        Assert.Equal("pulling manifest", ev.Status);
        Assert.Null(ev.Total);
        Assert.Null(ev.Completed);
        Assert.Null(ev.Digest);
    }

    // OllamaPullEventTests.testDecodeProgressEvent
    [Fact]
    public void DecodeProgressEventComputesFraction()
    {
        const string json = """
        {"status":"downloading","digest":"sha256:abc","total":2019377664,"completed":1000000}
        """;
        var ev = JsonSerializer.Deserialize<OllamaPullEvent>(json, Opts)!;
        Assert.Equal("downloading", ev.Status);
        Assert.Equal("sha256:abc", ev.Digest);
        Assert.Equal(2019377664L, ev.Total);
        Assert.Equal(1000000L, ev.Completed);
        Assert.NotNull(ev.FractionComplete);
        Assert.Equal(1000000.0 / 2019377664.0, ev.FractionComplete!.Value, 9);
    }

    // OllamaPullEventTests.testFractionCompleteIsNilWhenMissing
    [Fact]
    public void FractionCompleteNullWhenFieldsMissing()
    {
        var ev = JsonSerializer.Deserialize<OllamaPullEvent>("""{"status":"verifying"}""", Opts)!;
        Assert.Null(ev.FractionComplete);
    }

    // OllamaPullEventTests.testIsTerminalSuccess
    [Fact]
    public void IsTerminalSuccessOnlyForSuccessStatus()
    {
        Assert.True(new OllamaPullEvent("success").IsTerminalSuccess);
        Assert.False(new OllamaPullEvent("downloading").IsTerminalSuccess);
    }
}
